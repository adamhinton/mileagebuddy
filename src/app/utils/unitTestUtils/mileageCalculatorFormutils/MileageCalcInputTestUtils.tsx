// README
// This is utils for testing the MileageCalcFormTextInput and MileageCalcFormNumInput components
// They're very similar, so share much utils logic
// This is for mocking relevant functions, and creating a wrapper component for testing

import MileageCalcFormNumInput from "@/app/calculator/CalculatorFormComponents/MileageCalcFormNumberInput";
import MileageCalcFormTextInput from "@/app/calculator/CalculatorFormComponents/MileageCalcFormTextInput";
import { Path } from "react-hook-form";
import { z } from "zod";
import { Vehicle_For_db_POST } from "../../server/types/VehicleTypes/POSTVehicleTypes";

const mockRegister = jest.fn().mockReturnValue({
	onChange: jest.fn(),
	onBlur: jest.fn(),
	name: "test",
});

export const MileageInputNumTestWrapper = ({
	path,
	error,
	schema,
}: {
	path: Path<Vehicle_For_db_POST>;
	error?: string;
	schema: z.ZodNumber;
}) => {
	return (
		<MileageCalcFormNumInput
			registerFn={mockRegister}
			path={path}
			error={error}
			subSchema={schema}
		/>
	);
};

export const MileageInputTextTestWrapper = ({
	path,
	error,
	schema,
}: {
	path: Path<Vehicle_For_db_POST>;
	error?: string;
	schema: z.ZodString;
}) => {
	return (
		<MileageCalcFormTextInput
			registerFn={mockRegister}
			path={path}
			error={error}
			subSchema={schema}
		/>
	);
};
