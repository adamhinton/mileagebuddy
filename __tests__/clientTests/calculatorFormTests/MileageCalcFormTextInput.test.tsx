// README
// This is test for the MileageCalcFormTextInput component (obviously)
// It shares a lot of logic with MileageCalcFormNumInput, so these two tests share a utils file and many of the tests will look similar

import { MileageInputTextTestWrapper } from "@/app/utils/unitTestUtils/mileageCalculatorFormutils/MileageCalcInputTestUtils";
import { render, screen } from "@testing-library/react";
// Importing this is essential for stuff like toBeVisible()
import "@testing-library/jest-dom";
import { z } from "zod";

describe("Sanity check", () => {
	it("should pass", () => {
		expect(1 + 1).toBe(2);
	});
});

// This test is run for both components
describe("Common Input Behavior", () => {
	it("Renders without errors", () => {
		render(
			<MileageInputTextTestWrapper
				path="vehicleData.vehicleName"
				schema={z.string().describe("My Vehicle Name")}
			/>
		);
		const label = screen.getByText("My Vehicle Name");
		const input = screen.getByTestId("vehicleData.vehicleName-input");

		expect(input).toBeVisible();
		expect(label).toBeVisible();
	});
	// it("renders with label from schema description", () => {});
	// it("renders with label from path when no description", () => {});
	// it("shows required asterisk for required fields", () => {});
	// it("does not show required asterisk for optional fields", () => {});
	// it("displays error message when error prop provided", () => {});
	// it("hides error message when error prop is undefined", () => {});
	// it("Toggles error message when re-rendered with different error prop", () => {});
});

describe("Text Input Specific", () => {
	// it("enforces maxLength from schema", () => {});
	// it("enforces minLength from schema", () => {});
	// it("accepts valid text input", () => {});
});
