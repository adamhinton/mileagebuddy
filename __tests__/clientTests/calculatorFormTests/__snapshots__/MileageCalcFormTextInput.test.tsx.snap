// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Tests matches snapshot for basic text input from 4.30.2025 1`] = `
<div>
  <div
    class="mb-4"
  >
    <label
      class="mileage-calc-form-label"
      data-testid="vehicleData.vehicleName-label"
      for="vehicleData.vehicleName"
    >
      My Vehicle Name
      <span
        class="mileage-calc-form-required-asterisk"
      >
        *
      </span>
    </label>
    <input
      class="mileage-calc-form-text-field "
      data-testid="vehicleData.vehicleName-input"
      id="vehicleData.vehicleName"
      name="test"
      required=""
      type="text"
    />
  </div>
</div>
`;

exports[`Snapshot Tests matches snapshot for required text field from 4.30.2025 1`] = `
<div>
  <div
    class="mb-4"
  >
    <label
      class="mileage-calc-form-label"
      data-testid="vehicleData.vehicleName-label"
      for="vehicleData.vehicleName"
    >
      vehicleName
      <span
        class="mileage-calc-form-required-asterisk"
      >
        *
      </span>
    </label>
    <input
      class="mileage-calc-form-text-field "
      data-testid="vehicleData.vehicleName-input"
      id="vehicleData.vehicleName"
      minlength="1"
      name="test"
      required=""
      type="text"
    />
  </div>
</div>
`;

exports[`Snapshot Tests matches snapshot for text input with error from 4.30.2025 1`] = `
<div>
  <div
    class="mb-4"
  >
    <label
      class="mileage-calc-form-label"
      data-testid="vehicleData.vehicleName-label"
      for="vehicleData.vehicleName"
    >
      vehicleName
      <span
        class="mileage-calc-form-required-asterisk"
      >
        *
      </span>
    </label>
    <input
      aria-describedby="vehicleData.vehicleName-error"
      class="mileage-calc-form-text-field mileage-calc-form-text-field-error"
      data-testid="vehicleData.vehicleName-input"
      id="vehicleData.vehicleName"
      name="test"
      required=""
      type="text"
    />
    <p
      class="mileage-calc-form-error-message"
      data-testid="vehicleData.vehicleName-error"
      id="vehicleData.vehicleName-error"
    >
      <svg
        class="mileage-calc-form-error-icon"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10 18a8 8 0 100-16 8 8 0 000 16zM8.28 7.22a.75.75 0 00-1.06 1.06L8.94 10l-1.72 1.72a.75.75 0 101.06 1.06L10 11.06l1.72 1.72a.75.75 0 101.06-1.06L11.06 10l1.72-1.72a.75.75 0 00-1.06-1.06L10 8.94 8.28 7.22z"
          fill-rule="evenodd"
        />
      </svg>
      This is an error message
    </p>
  </div>
</div>
`;

exports[`Snapshot Tests matches snapshot for text input with length constraints from 4.30.2025 1`] = `
<div>
  <div
    class="mb-4"
  >
    <label
      class="mileage-calc-form-label"
      data-testid="vehicleData.vehicleName-label"
      for="vehicleData.vehicleName"
    >
      vehicleName
      <span
        class="mileage-calc-form-required-asterisk"
      >
        *
      </span>
    </label>
    <input
      class="mileage-calc-form-text-field "
      data-testid="vehicleData.vehicleName-input"
      id="vehicleData.vehicleName"
      maxlength="20"
      minlength="5"
      name="test"
      required=""
      type="text"
    />
  </div>
</div>
`;
